Aggregation step skew recovery: restart from beginning. (#616)
ietf-wg-ppm/draft-ietf-ppm-dap: draft-ietf-ppm-dap.md

body MUST fail with an HTTP client error status code. For further requests with
the same `AggregationJobInitReq` in the body, the Helper SHOULD respond as it
did for the original `AggregationJobInitReq`, or otherwise fail with an HTTP
client error status code. The Helper SHOULD allow the aggregation job to be
restarted even if the aggregation job has been continued (see

> {{agg-continue-flow}}), to allow for a Leader to recover from state skew between

Reviewer: Restarting entails re-sending the `AggregationJobInitReq` _unchanged_, correct?

Author: Yes, this is spelled out in the "Recovering from Aggregation Step Skew" section; I changed the wording to spell it out here, too.

> client error status code. The Helper SHOULD allow the aggregation job to be
> restarted even if the aggregation job has been continued (see
> {{agg-continue-flow}}), to allow for a Leader to recover from state skew between
> the Leader and Helper by restarting the aggregation; see

Reviewer 2: Mentioning restarting the aggregation twice is redundant.

```suggestion
client error status code. The Helper SHOULD allow the aggregation job to be
restarted even if the aggregation job has been continued (see
{{agg-continue-flow}}) to allow a Leader to recover from state skew; see
```

> {{aggregation-step-skew-recovery}} for more detail on the recovery process.

Reviewer: I think there's terminology here that we need to define more explicitly:
* "skew": Perhaps add a sentence like `The state of an aggregation job is said to be "skewed" if ...`
* "replay": I actually think we should try and avoid overloading this term (it's already used in the context of overexposing reports).

Author: * "skew": Done. While wordsmithing I also added the exact condition that the Leader would re-send the original request, i.e. on receiving a `stepMismatch` error.
* "replay": makes sense, I replaced these usages of "replay" with "restart".

#### Recovering from Aggregation Step Skew {#aggregation-step-skew-recovery}

An aggregation job may require many round trips over the network to complete. If
one of the Aggregators fails during this process in a manner that results in it
losing its state, then the Aggregators risk losing the data associated with the

> aggregation job, or in some cases all of the batches associated with the

Reviewer 2: ```suggestion
aggregation job, or in some cases, all of the batches associated with the
```

> aggregation job (if one of the Aggregators considers the aggregation to be
> complete but the other does not, the Aggregators will encounter a
> `batchMismatch` error on attempting to collect the batches associated with the
> aggregation job). This section describes how Aggregators can recover from this

Reviewer: ```suggestion
aggregation job. (For example, if one of the Aggregators considers the aggregation to be
complete but the other does not, the Aggregators will encounter a
`batchMismatch` error on attempting to collect the batches associated with the
aggregation job.) This section describes how Aggregators can recover from this
```

> aggregation job). This section describes how Aggregators can recover from this
> scenario.

Reviewer: ```suggestion
aggregation job). This section describes how Aggregators can recover from such
failures.
```

`AggregationJobContinueReq` messages contain a `step` field, allowing
Aggregators to ensure that their peer is on the expected step of DAP
aggregation. An aggregation job is said to be "skewed" if the Leader and Helper
have a different view of the last completed step of aggregation; for example, if

> the Leader does not receive or successfully process an `AggregationJobResp` from

Reviewer: "say, due to a failure of the underlying transport"

the Helper, the Leader will believe that the last completed step is one less
than the Helper. The intent of the `step` field is to allow detection and
recovery from a skewed aggregation job. When the Leader receives a
`stepMismatch` error, it SHOULD re-send the original `AggregationJobInitReq`
message to restart the aggregation process and allow aggregation to complete

> successfully. To make this kind of recovery possible, Leader implementations
> SHOULD retain enough information to reconstruct the `AggregationJobInitReq`
> message in the case that the aggregation process becomes desynchronized.

Reviewer: I think this SHOULD is redundant, and maybe can be cut completely? What's the intent of this sentence?
```suggestion
successfully.
```

Author: I'm explicitly suggesting/hinting to Leader implementations that they need to keep the original request around (or enough information to reconstruct it byte-for-byte), if they want to implement skew recovery.

I suppose this is implied well enough by the previous sentence, and the following description of the Helper behavior -- I removed this sentence.

Changing an aggregation job's parameters is illegal, so the Helper MUST ensure
that the Leader's `AggregationJobInitReq` message did not change when it was
re-sent (i.e., the two messages must be identical). This prevents the Leader
from re-winding an aggregation job and re-running aggregation with different
parameters.

One way the Helper could implement this would be to store a digest of the
Leader's request, indexed by aggregation job ID, and refuse to service a re-sent

> request unless it matches the previously seen request.

Reviewer: FWIW each AggreagtionJobContinueReq needs to be the same as well. Though I believe the protocol logic should ensure this? It's still worth mentioning in any case.

Also the responses shouldn't change either.

Reviewer 2: +1. The Leader could restart an aggregation job and arbitrarily remove valid report shares from subsequent AggregationJobContinueReqs. In that case, the Helper may need to ensure the request digest hasn't changed for every single step involved.

It could do this by storing the digest indexed by aggregation job ID and step. But, this increases the amount of state required from one digest (in the -1 recovery) to n digests (in from-the-beginning recovery).

Author: Can y'all spell out the specific problem or attack you have in mind?

I don't think we need to ensure the later requests are the same. My reasoning is:
* If the Helper did not consider the aggregation process complete when the retry was initiated, no state outside of the aggregation job has been modified -- so if the Leader ends up removing a different set of reports on retry, there is no problem.
* If the Helper did consider the aggregation process complete when the retry was initiated, it will have already committed the recovered output shares & noted that it has done so, and thus will not commit a second time. If the Leader aggregates a different set of reports, this will result in a `batchMismatch` error, but the Leader has no ability to e.g. aggregate a slightly different set of reports to infringe on a particular report's privacy.

OTOH, the Leader _accidentally_ aggregating a different set of reports against an aggregation job the Helper considers to be complete is likely a downside of this approach -- it would be hard to avoid in corner cases. The most likely specified corner cases I can find here would be (a) the HPKE key rotating or (b) the replay-checking state being dropped, both of which would result in the report becoming un-aggregatable on retry. (Tracking the exact request/responses at each step wouldn't help -- we'd fail earlier & maybe with a better error code, but would still eventually encounter a `batchMismatch` error on collection, since the Leader would never be able to complete the aggregation job. I think a truly robust implementation of the restart-from-beginning idea would require the Helper to have the ability to un-commit an aggregation job from the running-total aggregate shares when the aggregation job is restarted, which I suppose would require tracking the aggregate shares for each aggregation job -- i.e. we'd lose the storage savings of the running-total aggregation strategy.)

Reviewer: Good point. I was originally thinking of something like this: the Helper remembers a hash of each request, and if the request doesn't match the hash, then it aborts the aggregation job. But you're right that we need to be robust to either the Leader or Helper making different rejection decisions.

Another wrinkle pointed out by @divergentdave on the 2024/10/23 on the call: we need to be careful about VDAF preparation. Existing security analysis assumes the attacker can't "rewind" the computation and retry with different inputs. In particular, we're worried that reusing the Prio3 proof shares may result in a privacy violation. (Likewise for the Beaver triples in Poplar1.) In particular we need to ensure the Leader doesn't change the prep share it sends for each report.

Author: There was some off-PR discussion about this today, written up in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/604#issuecomment-2433065034.

The summary is that not only do we need the Helper to verify that the entire sequence of requests is the same on retry, we _also_ need the Leader to verify that the entire sequence of responses is the same on retry.

This is achieved by the restart-from-previous-step currently specified in DAP-12, but is not achieved by the restart-from-beginning (this PR) or restart-from-anywhere (#569) strategies as currently written. We could certainly augment this PR or #569 to specify the requirement that the Leader record the sequence of responses it has successfully processed; but AIUI the original motivation for changing the retry strategy was to permit a Leader implementation which did not need to record anything to durable storage between aggregation steps. If the Leader now has to store response hashes to durable storage between aggregation steps, we no longer meet the original motivation for modifying the retry strategy.

Reviewer 2: Thank you for the detailed write up and I agree. Although this strategy would work for two happy Aggregators, it cannot protect against scenarios where one is malicious, compromising our security guarantees. If the Leader and Helper would need to remember and enforce the exact digests for each step anyways, we may as well just restart from the last "committed" step between the two, which falls into line with the current status quo.

Implementation note: in the case that the Leader restarts an aggregation which

> the Helper considers to be complete, the Helper must ensure it does not commit

Reviewer: ```suggestion
the Helper considers to be complete, the Helper MUST NOT commit
```

Author: Done. Is it appropriate to include a `MUST` in an implementation note? (If not, I think this behavior is important enough to promote from an implementation note into regular text.)

Reviewer: Yes, I think it's appropriate.

#### Recovering from Aggregation Step Skew {#aggregation-step-skew-recovery}

`AggregationJobContinueReq` messages contain a `step` field, allowing
Aggregators to ensure that their peer is on the expected step of DAP
aggregation. In particular, the intent is to allow detection and recovery from a
scenario where the Helper and the Leader have differing views of the current
aggregation step, which may occur if an `AggregationJobResp` response to the
Leader is dropped due to a transient network failure. The Leader could then
resend the original `AggregationJobInitReq` message to restart the aggregation
process, allowing aggregation to complete successfully. To make this kind of
recovery possible, Leader implementations SHOULD retain enough information to
reconstruct the `AggregationJobInitReq` message in the case that the aggregation
process becomes desynchronized.

> When implementing an aggregation step skew recovery strategy, the Helper SHOULD

Reviewer: Here we should repeat the line about "changing an aggregation job's parameters is illegal".

Author: OK. Perhaps more controversially, I changed the `SHOULD` to a `MUST` to match the `MUST` around "changing an aggregation job's parameters is illegal" -- I think these were effectively in conflict in terms of how required this was.

(Historically, I think we specified a `SHOULD` here to permit an implementation to ignore the request body on a re-send -- other than the `step` parameter, I suppose -- and blindly reply with the previous response. I'm OK dropping this because (a) no one has expressed interest in implementing this and (b) boy howdy, those semantics would be challenging to debug -- the response would potentially be completely independent of the request.)

ensure that the Leader's `AggregationJobInitReq` message did not change when it
was re-sent (i.e., the two messages must be identical). This prevents the Leader
from re-winding an aggregation job and re-running aggregation with different
parameters.

One way the Helper could implement this would be to store a digest of the
Leader's request, indexed by aggregation job ID, and refuse to service a re-sent
request unless it matches the previously seen request.

Implementation note: in the case that the Leader replays an aggregation which

> the Helper considers to be complete, the Helper should ensure it does not commit

Reviewer: ```suggestion
the Helper considers to be complete, the Helper must ensure it does not commit
```
